### PR TITLE
Spruce- Ivette F.

### PR DESCRIPTION
**Pull request by l1zardmcguire**

### CODE REVIEW

File: swap_meet/clothing.py [hunk 1]
```diff
@@ -1,2 +1,9 @@
-class Clothing:
-    pass
+from swap_meet.item import Item
+
+class Clothing(Item):
+    def __init__(self, category="Clothing", condition=0, age=None):
+        super().__init__(category, condition, age)
```
Comment (anselrognlie) [inline swap_meet/clothing.py:5]: We don't really want to let users of our  class change the category for a Clothing instance to something else (at least not from the constructor). We can leave the category out of the constructor parameters entirely, and pass the desired category literally in to the parent's constructor as follows:
```python
    def __init__(self, condition = 0, age = None):
        super().__init__("Clothing", condition, age)
```

File: swap_meet/decor.py [hunk 2]
```diff
@@ -1,2 +1,9 @@
-class Decor:
-    pass
+from swap_meet.item import Item
+
+class Decor(Item):
+    def __init__(self, category="Decor", condition=0, age=None):
+        super().__init__(category, condition, age)
```
Comment (anselrognlie) [inline swap_meet/decor.py:5]: The same situation applies here as for the Clothing constructor.

File: swap_meet/electronics.py [hunk 3]
```diff
@@ -1,2 +1,8 @@
-class Electronics:
-    pass
+from swap_meet.item import Item
+
+class Electronics(Item):
+    def __init__(self, category="Electronics", condition=0, age=None):
+        super().__init__(category, condition, age)
```
Comment (anselrognlie) [inline swap_meet/electronics.py:5]: The same situation applies here as for the Clothing constructor.

File: swap_meet/item.py [hunk 4]
```diff
@@ -1,2 +1,23 @@
 class Item:
-    pass
+    def __init__(self, category="", condition=0, age=None):
```
Comment (anselrognlie) [inline swap_meet/item.py:2]: I was wondering why you decided to use `None` as the default for age but I see that it looks like you decided to exclude any items from the age-related extension features unless it was explicitly set. Neat decision!

File: swap_meet/item.py [hunk 5]
```diff
@@ -1,2 +1,23 @@
...
+        self.category = category
+        self.condition = condition
+        self.age = age
+
+    def __str__(self):
+        return "Hello World!"
+
+    def condition_description(self):
```
Comment (anselrognlie) [inline swap_meet/item.py:10]: What would happen if we tried to get the description for an Item with a condition of 3.5 (there are tests that set the condition to 3.5, but don't try to read the description). The description test checks for a condition of 1 and a description of 5, but we should still make sure our code covers any other reasonable conditions, including floating point values (since we've seen them used in other tests).

How could we modify this (slightly) to cover the values between the whole numbers?

File: swap_meet/vendor.py [hunk 6]
```diff
@@ -1,2 +1,97 @@
+from swap_meet.item import Item
+
 class Vendor:
-    pass
+    def __init__(self, inventory=None):
+        self.inventory = inventory if inventory is not None else []
```
Comment (anselrognlie) [inline swap_meet/vendor.py:5]: Nice job handling the default case for an empty inventory.

File: swap_meet/vendor.py [hunk 7]
```diff
@@ -1,2 +1,97 @@
...
+    def add(self, item):
+        self.inventory.append(item)
+        return item
+
+    def remove(self, item):
+        if item in self.inventory:
+            self.inventory.remove(item)
+            return item
+        else:
+            return False
```
Comment (anselrognlie) [inline swap_meet/vendor.py:16]: I like reversing the sense of the condition to be able to treat it liken a guard clause, which allows us to outdent the main logic of the function something like the following:
```python
        if item not in self.inventory:
            return False

        self.inventory.remove(item)
        return item
```

Another approach we could take is to try to remove the item directly, and handle the `ValueError` that occurs if it's not there, and return `False` to handle it, similar to the commented out code you have for `swap_items`.

File: swap_meet/vendor.py [hunk 8]
```diff
@@ -1,2 +1,97 @@
...
+
+    def get_by_category(self, category):
+        items_in_category = []
+        for item in self.inventory:
+            if item.category ==  category:
+                items_in_category.append(item)
```
Comment (anselrognlie) [inline swap_meet/vendor.py:22]: Notice this pattern of 
```python
result_list = []
for element in source_list:
    if some_condition(element):
        result_list.append(element)
```

can be rewritten as a list comprehension as
```python
result_list = [element for element in source_list if some_condition(element)]
```

Which here would look like
```python
        items_in_category = [item for item in self.inventory if item.category == category]
```

File: swap_meet/vendor.py [hunk 9]
```diff
@@ -1,2 +1,97 @@
...
+        return items_in_category
+
+    def swap_items(self, vendor, my_item, their_item):
+        # try:
```
Comment (anselrognlie) [inline swap_meet/vendor.py:26]: In general, this is a reasonable approach, but notice how needing to restore the first list if the second remove fails adds to the complexity. And ideally, we would like to have `my_item` end up back in its inventory in the original position it started, as though no operation had been attempted. Overall, for this operation, I'd say that the approach you left in place is the more maintainable approach.

File: swap_meet/vendor.py [hunk 10]
```diff
@@ -1,2 +1,97 @@
...
+        #     vendor.inventory.remove(their_item)
+        #     self.inventory.append(their_item)
+        # except ValueError:
+        #     self.inventory.append(my_item)
+        #     vendor.inventory.remove(my_item)
+        #     return False
+        # return True
+
+        if my_item in self.inventory and their_item in vendor.inventory:
```
Comment (anselrognlie) [inline swap_meet/vendor.py:40]: As with the remove method, I'd tend to invert the condition here to use it as a guard clause. The same goes for any of the subsequent methods where there's a condition with an `else` which only has a `return` in it.

Nice reuse of the add and remove helpers!

File: swap_meet/vendor.py [hunk 11]
```diff
@@ -1,2 +1,97 @@
...
+            self.remove(my_item)
+            vendor.add(my_item)
+            vendor.remove(their_item)
+            return True
+        else:
+            return False
+
+    def swap_first_item(self, vendor):
+        if self.inventory and vendor.inventory:
+            return self.swap_items(vendor, self.inventory[0], vendor.inventory[0])
```
Comment (anselrognlie) [inline swap_meet/vendor.py:51]: Nice reuse of `swap_items`.

Notice, though, that by using `swap_items`, this winds up being O(n) (mostly due to removing the items from the inventory). We could simply swap the items in place here to have O(1) runtime with something like:
```python
        self.inventory[0], vendor.inventory[0] = vendor.inventory[0], self.inventory[0]
```

Alternatively, we could think about how to modify the implementation of `swap_items` so that it would still work in general, but also give us O(1) in this specific case. Try that out if that sounds interesting!

File: swap_meet/vendor.py [hunk 12]
```diff
@@ -1,2 +1,97 @@
...
+        else:
+            return False
+    
+    def get_best_by_category(self, category):
+        matches_category = self.get_by_category(category)
+        best_item = None
+        if matches_category:
+            best_item = max(matches_category, key = lambda item : item.condition)
+        return best_item
```
Comment (anselrognlie) [inline swap_meet/vendor.py:60]: `max` raises a `ValueError` is the list is empty. Doing an explicit check for an empty list will protect us from that. But `max` also has a `default` parameter that we can use to tell `max` what to return in case of an empty list. That would look something like:
```python
        return max(matches_category, key = lambda item : item.condition, default=None)
```

File: swap_meet/vendor.py [hunk 13]
```diff
@@ -1,2 +1,97 @@
...
+
+    def swap_best_by_category(self, other, my_priority, their_priority):
+        my_best_item = self.get_best_by_category(their_priority)
+        their_best_item = other.get_best_by_category(my_priority)
+        if my_best_item and their_best_item:
+            self.swap_items(other, my_best_item, their_best_item)
+            return True
+        else:
+            return False
```
Comment (anselrognlie) [inline swap_meet/vendor.py:69]: Because of how `swap_items` is implemented, the empty checks here aren't necessary. So long as we make use of the return value from `swap_items`, we could get away with
```python
        return self.swap_items(other, my_best_item, their_best_item)
```

File: swap_meet/vendor.py [hunk 14]
```diff
@@ -1,2 +1,97 @@
...
+
+    def get_by_newest(self):
+        """
+        returns the newest item in inventory; if multiple newest items, returns the first one listed in the inventory
+        """
+        only_ages = []
+        for item in self.inventory:
+            if item.age != None:
+                only_ages.append(item)
```
Comment (anselrognlie) [inline swap_meet/vendor.py:78]: Notice the list comprehension pattern here. From the earlier comment, how could we make this into a list comprehension?

File: swap_meet/vendor.py [hunk 15]
```diff
@@ -1,2 +1,97 @@
...
+        if only_ages:
+            newest_item = min(only_ages, key = lambda item : item.age)
```
Comment (anselrognlie) [inline swap_meet/vendor.py:80]: Like `max`, `min` also supports a default parameter. How would that let us simplify this?

File: swap_meet/vendor.py [hunk 16]
```diff
@@ -1,2 +1,97 @@
...
+            return newest_item
+        else:
+            return None
+
+    def swap_by_newest(self, other):
+        """
+        
+        """
+        my_newest = self.get_by_newest()
```
Comment (anselrognlie) [inline swap_meet/vendor.py:89]: Nice use of the extra helper function you added!

File: swap_meet/vendor.py [hunk 17]
```diff
@@ -1,2 +1,97 @@
...
+        their_newest = other.get_by_newest()
+
+        if my_newest and their_newest:
+            self.swap_items(other, my_newest, their_newest)
+            return True
+        else: 
+            return False
```
Comment (anselrognlie) [inline swap_meet/vendor.py:96]: Notice how similar this is to `swap_best_by_category`. Knowing how `swap_items` operates, how could we simplify this part?

File: tests/test_enhancements.py [hunk 18]
```diff
@@ -0,0 +1,122 @@
+import pytest
+from swap_meet.vendor import Vendor
+from swap_meet.clothing import Clothing
+from swap_meet.decor import Decor
+from swap_meet.electronics import Electronics
+
+def test_get_by_newest():
```
Comment (anselrognlie) [inline tests/test_enhancements.py:7]: Nice job adding tests for your age-related functions. One thing I'd be a little careful of is that for most of these tests, the newest item happens to be the first thing in the inventory. Try having the newest at a more varied distribution of locations. Until the very final test, I think most of these tests would pass if `get_newest_item` simply returned the first thing in the inventory.

File: tests/test_enhancements.py [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+    tai = Vendor(
+        inventory=[item_a, item_b, item_c, item_d, item_e]
+    )
+
+    newest_item = tai.get_by_newest()
+
+    assert newest_item.category == "Clothing"
+    assert newest_item.age == pytest.approx(2.0)
+
+def test_get_by_newest_some_items_with_None_age():
```
Comment (anselrognlie) [inline tests/test_enhancements.py:22]: This test doesn't really prove to me that this is ignoring the None values. Again, from this, it could just be that the first item is being returned immediately and the items with `None` for their age aren't even being examined. We could make this a more robust test by moving the newest item so that it came between to `None` records (it would then be reasonable that the code had to skip over at least _one_ of them). Or have a test where we try to get the newest item from a list containing _only_ items with `None` ages, which should give us `None` as a result.

**Review by anselrognlie — COMMENTED** [review]
Nice job!

All required waves are complete and all required tests (and then some) are passing! Your code is well organized, and you are generally using good, descriptive variable names. Your general approaches are good, you're off to a good start with OOP, and you're making good reuse of functions.

The main things I would recommend focusing are are in some of the smaller details. For instance, notice that none of the Item child classes have tests that need a category parameter. And in fact, we would probably want to prevent a user from creating an Electronics with a category of "Decor", so really consider which constructor parameters we need to provide.

Also, be sure to be reading the tests carefully to understand what they're testing, as well as what they're not. Be sure to think about other situations that our code might need to handle (especially situations that show up in other tests, but which might not themselves be tested). The supplied tests are an important baseline, but we should still think about the overall behavior of our code and how it handles a variety of situations.

And nice job adding the age property and swap method! Adding the helper method worked really well, and also made it easier to add more targeted unit tests.

Well done!
